go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +631,562 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+}
+
+// ApplySettings implements pool.Resource.
+func (vbc *vcopierCopyWorker) ApplySettings(context.Context, []string) error {
+	return nil
+}
+
+// Execute advances a task through each state until it is canceled, completed
+// or failed.
+func (vbc *vcopierCopyWorker) Execute(ctx context.Context, task *vcopierCopyTask) *vcopierCopyTaskResult {
+	startedAt := time.Now()
+	state := vcopierCopyTaskPending
+
+	var err error
+
+	// As long as the current state is not done, keep trying to advance to the
+	// next state.
+	for nextState := state.Next(); !state.IsDone(); nextState = state.Next() {
+		var advanceFn func(context.Context, *vcopierCopyTaskArgs) error
+
+		// Get the advanceFn to use to advance the task to the nextState.
+		switch nextState {
+		case vcopierCopyTaskBegin:
+			advanceFn = func(context.Context, *vcopierCopyTaskArgs) error {
+				// Rollback to make sure we're in a clean state.
+				if err := vbc.vdbClient.Rollback(); err != nil {
+					return fmt.Errorf("failed to rollback: %s", err.Error())
+				}
+				// Begin transaction.
+				if err := vbc.vdbClient.Begin(); err != nil {
+					return fmt.Errorf("failed to start transaction: %s", err.Error())
+				}
+				return nil
+			}
+		case vcopierCopyTaskInsertRows:
+			advanceFn = func(ctx context.Context, args *vcopierCopyTaskArgs) error {
+				if _, err := vbc.insertRows(ctx, args.Rows()); err != nil {
+					return fmt.Errorf("failed inserting rows: %s", err.Error())
+				}
+				return nil
+			}
+		case vcopierCopyTaskUpdateCopyState:
+			advanceFn = func(ctx context.Context, args *vcopierCopyTaskArgs) error {
+				if err := vbc.updateCopyState(ctx, args.Lastpk()); err != nil {
+					return fmt.Errorf("error updating _vt.copy_state: %s", err.Error())
+				}
+				return nil
+			}
+		case vcopierCopyTaskCommit:
+			advanceFn = func(context.Context, *vcopierCopyTaskArgs) error {
+				// Commit.
+				if err := vbc.vdbClient.Commit(); err != nil {
+					return fmt.Errorf("error commiting transaction: %s", err.Error())
+				}
+				return nil
+			}
+		case vcopierCopyTaskComplete:
+			advanceFn = func(context.Context, *vcopierCopyTaskArgs) error { return nil }
+		}
+
+		// Maybe we should panic here?
+		if advanceFn == nil {
+			err = fmt.Errorf("don't know how to advance from %s to %s", state, nextState)
+			state = vcopierCopyTaskFail
+			break
+		}
+
+		// TryAdvance tries to execute advanceFn, and also executes any
+		// lifecycle hooks surrounding the provided state.
+		//
+		// If all lifecycle hooks and advanceFn are successfully executed,
+		// TryAdvance will return state, which should be == the nextState that
+		// we provided.
+		//
+		// If there was a failure executing lifecycle hooks or advanceFn,
+		// TryAdvance will probably return a failure state (i.e. canceled or
+		// failed), along with a diagnostic error.
+		if state, err = task.Lifecycle().TryAdvance(ctx, task.Args(), nextState, advanceFn); err != nil {
+			break
+		}
```

Comment:
I'm a little uneasy/uncertain about the error/failure handling here. We can walk through it.